data/completion/etelpmoc.sh

```diff
@@ -0,0 +1,180 @@
+    \compgen "$@"
+}
+alias compgen=xcompgen
+shopt -s expand_aliases
+
+# compopt replaces the original compopt with one that just sets/unsets entries
+# in _compopts
+compopt() (
+    local i
+
+    for ((i=0; i<$#; i++)); do
+        case "${!i}" in
+            -o)
+                ((i++))
+                _compopts[${!i}]=1
+                ;;
+            +o)
+                ((i++))
+                unset _compopts[${!i}]
+                ;;
+        esac
+    done
+)
+
+_compfunc="_minimal"
+_compact=""
+# this is a lot more complicated than it should be, but it's how you
+# get the result of 'complete -p "$1"' into an array, splitting it as
+# the shell would.
+readarray -t _comp < <(xargs -n1 < <(complete -p "$1") )
+if [[ "${_comp[*]}" ]]; then
+    while getopts :abcdefgjksuvA:C:W:o:F: opt "${_comp[@]:1}"; do
+        case "$opt" in
+            a)
+                _compact="alias"
+                ;;
+            b)
+                _compact="builtin"
+                ;;
+            c)
+                _compact="command"
+                ;;
+            d)
+                _compact="directory"
+                ;;
+            e)
+                _compact="export"
+                ;;
+            f)
+                _compact="file"
+                ;;
+            g)
+                _compact="group"
+                ;;
+            j)
+                _compact="job"
+                ;;
+            k)
+                _compact="keyword"
+                ;;
+            s)
+                _compact="service"
+                ;;
+            u)
+                _compact="user"
+                ;;
+            v)
+                _compact="variable"
+                ;;
+            A)
+                _compact="$OPTARG"
+                ;;
+            o)
+                _compopts["$OPTARG"]=1
+                ;;
+            C|F)
+                _compfunc="$OPTARG"
+                ;;
+            W)
+                readarray -t COMPREPLY < <( \compgen -W "$OPTARG" -- "${COMP_WORDS[$COMP_CWORD]}" )
+                _compfunc=""
+                ;;
+            *)
+                # P, G, S, and X are not supported yet
+                _die "ERROR: unknown option -$OPTARG"
+                ;;
+        esac
+    done
+fi
+
+_bounce=""
+case "$_compact" in
+    # these are for completing things that'll be interpreted by the
+    # "outside" bash, so send them back to be completed there.
+    "alias"|"export"|"job"|"variable")
+        _bounce="$_compact"
+        ;;
+esac
+
+if [ ! "$_bounce" ]; then
+    if [ "$_compact" ]; then
+        readarray -t COMPREPLY < <( \compgen -A "$_compact" -- "${COMP_WORDS[$COMP_CWORD]}" )
+    elif [ "$_compfunc" ]; then
+        # execute completion function (or the command if -C)
+        $_compfunc
+    fi
+fi
+
+# print completions to stdout
+echo "${!_compopts[@]}"
+echo "$_bounce"
+echo ""
+printf "%s\n" "${COMPREPLY[@]}"
```

Comment:
As discussed on IRC, what I'm thinking about here is the transition from confined to unconfined. Ie, everything up until `snap run command=complete` is fine because it is running within the user's context with only trusted input and code. Everything after that and until etelpmoc.sh exits is ok and no worse than running non-bash-completion snap cli commands in the shell because it is running under the confinement of the snap. etelpmoc.sh then fills various variables for the unconfined shell to interpret based on the completer script in the snap and the unconfined bash displays this to the user-- it is this handoff I am concerned about.

More specifically, I think wrt this transition that there are two things to think about:
1. that a malicious completer script can feed something back to etelpmoc.sh and ultimately bash that causes bash to misbehave. This would arguably be a bug in bash, but perhaps there is output validation from etelpmoc.sh that can be done to put guard rails on potential issues
2. tricking the user into completing something that she didn't want to, but accidentally did

I've been looking at '2' this afternoon (I'll continue to look at '1' later) and found there are a number of ways to trick the user with just ';':

```
$ mkdir bar
$ cd bar
$ complexion <tab>
  |
   -> $ complexion ;ls ~/.ssh # user is one 'Enter' away from getting pwn'd
      $ ^C
$ touch 1 2
$ complexion <tab>
  | 
   -> $ complexion 
      1;ls ~/.ssh  2;ls ~/.ssh
      $ complexion 1<tab> # after typing '1<tab>, user is one 'Enter' away from getting pwn'd
      $ ^C
$ complexion <tab>
  |
   -> $ complexion
      1            2            3;ls ~/.ssh # note, '3' doesn't exist
      $ ^C
```

It is easy to imagine less obvious things, especially if doing command completion. Eg:
```
$ complexion <tab>
  |
   -> $ complexion
      current/1;/usr/bin/baz  current/7/usr/bin/foo   current/9/usr/bin/bar
      $ ^C
```

I'm sure creative folks could do a lot more. To me this clearly shows we need some sort of output validation in etelpmoc.sh such as piping through grep and disregarding any results that have weird stuff. Ideally this would be a whitelist, but I fear this is going to be problematic for filenames.... At an absolute minimum ';' needs to be removed, but really several things ('*' comes to mind, but likely others) are going to be potentially problematic. :\